### DISCUSSION TITLE
EV3 to PC Bluetooth Communication

### DISCUSSION THREAD

**Discussion by shoringit** [pybricks/support#1193] [Q&A] [answered]
The next thing I'm working on getting working properly under pybricks is Bluetooth communication.  I have it working EV3-to-EV3 just fine.  My next step is getting it to work EV3-to-PC, and I've hit a snag.  I've searched and haven't found a solution.

So, under the normal Lego EV3 firmware, if I pair the module to my computer, I get a serial connection port (SSP) on the windows side. This SSP port can be used with the built-in messaging blocks to send/receive info from a program on the PC (make a program that displays info from the EV3, or to receive commands from the PC for instance). 

Under pybricks, I'm not getting this port when I pair the EV3 to Windows.  It is just a limitation of the python EV3 image, or is there some way to get this functionality like I have with the original firmware?  Or is there another way to get this communication between the EV3 and the PC?

Thanks,

Danny

**Comment by laurensvalk** [comment, marked as answer]
These examples may be helpful:

https://github.com/pybricks/pybricks-api/tree/master/examples/ev3/bluetooth_pc

**Reply by shoringit** [reply]
Ok, never would have thought about this being implemented for the PC so it works the same as an EV3.  That's awesome!

Thanks,

Danny

**Reply by bradocchs** [reply]
This is awesome, but I'm new to Python and not getting this to work. I'm pretty sure it is not using the messaging partial implementation, so never even trying to connect. So far I've tried copying the pybricks folder that has the actual messaging and bluetooth code into my project, but python still uses the empty blueprint module instead. How do I get it to use the correct module? Thanks!

**Reply by bradocchs** [reply]
Quick update: I think it is now executing the correct module, but I'm getting a timeout error on the PC side in bluetooth.py socket.connect() method. I know I have the correct brick ID. I've also tried without pairing, paired/disconnected, and paired/connected. Same error. My PC is running Windows 11. Any troubleshooting hints appreciated. Thanks!

**Reply by dlech** [reply]
Can you give the exact steps to reproduce the problem, including the exact programs you ran on the EV3 and on the computer and the order in which you started each program and share the full text of the error message?

**Reply by bradocchs** [reply]
Step 1: Paired PC to EV3, but left disconnected (also tried connected before running)
Step 2: Download EV3 code to brick and run on brick
Step 3: Run PC code on Windows 11 using Python 3.11 (previously ran **pip install pybricks** and copied the **pybricks** directory that contains the implementation of **messaging.py** and **bluetooth.py** in my workspace - VS Code)

EV3:
```python
#!/usr/bin/env pybricks-micropython
from pybricks.hubs import EV3Brick
from pybricks.messaging import BluetoothMailboxServer, TextMailbox

ev3 = EV3Brick()
server = BluetoothMailboxServer()
mbox = TextMailbox('talk', server)
print("waiting for connection...")
server.wait_for_connection()
print("connected.")
while True:
    mbox.wait()
    ev3.speaker.say(mbox.read())
```
* this never prints "connected."

PC:
```python
from pybricks.messaging import BluetoothMailboxClient, TextMailbox
from time import sleep

client = BluetoothMailboxClient()
mbox = TextMailbox('talk', client)
client.connect("00:16:53:45:BA:79")
print("connected to brick 1...")
mbox.send("ready")

while True:
    sleep(1)
```
* this gives the timeout error:
```
Traceback (most recent call last):
  File "c:\Users\Brad\Documents\py-test1\test2.py", line 6, in <module>
    client.connect("00:16:53:45:BA:79")
  File "c:\Users\Brad\Documents\py-test1\pybricks\messaging.py", line 342, in connect
    client.handle_request()
  File "c:\Users\Brad\Documents\py-test1\pybricks\bluetooth.py", line 85, in handle_request
    self.socket.connect(self.client_address)
TimeoutError: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
```

Thanks again for taking a look at this!

Brad

PS - Not sure why my code is not formatting as code...

**Reply by dlech** [reply]
You are sure the Bluetooth address (`"00:16:53:XX:XX:XX"`) is the correct one for your EV3?

**Reply by dlech** [reply]
> PS - Not sure why my code is not formatting as code...

I fixed it up.  The trick is to use a "code fence with 3 backticks like this:

````
```python
# code goes here
...
```
````

**Reply by omegacoreFLL** [reply]
I have another issue when trying these steps. On the PC side it looks to be 'skipping' all the pybricks objects/methods. I've tried a bunch of ways to get a communication channel between the ev3 and my laptop, in the scope of transfering the position (odometry) from the main robot program to a pygame simulation, and the official messaging feature seems to be the way to go. 

I can't understand why can't I establish the connection: robot code would wait for connection infinitely, and the other program would print 'connected' instantly, not reading anything send from the host program. When reversing hosting, it's obviously similar.

The 'main.py' (ev3 code):

```#!/usr/bin/env pybricks-micropython

from pybricks.tools import wait, StopWatch
import math

from pybricks.messaging import *


.... imports here


server = BluetoothMailboxServer()

mbox = TextMailbox('greeting', server)


print('waiting for connection...')
server.wait_for_connection()
print('connected!')


mbox.wait()
print(mbox.read())
mbox.send('hello to you!')
```

Pygame code:

```
#!/usr/bin/env pybricks-micropython

from sim_constants import *

import pygame
import math

from pybricks.messaging import BluetoothMailboxClient, TextMailbox
from pybricks.tools import wait, StopWatch


client = BluetoothMailboxClient()

client.connect('ev3dev')
wait(5000) #doesn't work, like the other 'pybricks' imports

mbox = TextMailbox('greeting', client)

mbox.send('hello!')
mbox.wait()
print("Received message:", mbox.read())
```

Yes, I've installed pybricks on VS code, otherwise would get an error. I don't know what to do anymore.


Here are my robot's specifications:
```
========== ev3dev-sysinfo ==========
Image file:         ev3-micropython-v2.0.0-sd-card-image
Kernel version:     4.14.117-ev3dev-2.3.5-ev3
Brickman:           0.10.3
BogoMIPS:           148.88
Bluetooth:          4.0
Board:              board0
BOARD_INFO_HW_REV=7
BOARD_INFO_MODEL=LEGO MINDSTORMS EV3
BOARD_INFO_ROM_REV=6
BOARD_INFO_SERIAL_NUM=0016535181BD
BOARD_INFO_TYPE=main
```

**Comment by bradocchs** [comment]
> You are sure the Bluetooth address (`"00:16:53:XX:XX:XX"`) is the correct one for your EV3?

Is it different than the Brick ID/SN? I'm using the one listed at the bottom of the About page.

**Reply by bradocchs** [reply]
Never mind. It is different and I found the correct BT address and of course it works fine now. Thanks!

**Reply by bradocchs** [reply]
Ok. So using the PC as a client works great. However, when I tried using the PC as a server I get this error:
```
Traceback (most recent call last):
  File "c:\Users\Brad\Documents\talk\talk.py", line 4, in <module>
    server = BluetoothMailboxServer()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Brad\Documents\talk\pybricks\messaging.py", line 267, in __init__
    super(ThreadingRFCOMMServer, self).__init__(
  File "c:\Users\Brad\Documents\talk\pybricks\bluetooth.py", line 34, in __init__
    self.socket.bind(server_address)
PermissionError: [WinError 10013] An attempt was made to access a socket in a way forbidden by its access permissions
```
Any ideas?

**Reply by dlech** [reply]
Please see https://github.com/pybricks/support/issues/902#issuecomment-1374157195

**Reply by bradocchs** [reply]
I see. Modifying the RFCOMM channel is straightforward on the PC side. I would like to try to modify it on the EV3 if it's not too involved. When you said "We hard-coded EV3_RFCOMM_CHANNEL in messaging.py, so you would need to copy the definition of BluetoothMailboxServer and/or MailboxRFCOMMClient and modify it to take the channel as a parameter." does this mean you can make the change in python without having to recompile the entire pybricks module? Would you be willing to explain the steps a bit more or point me to resources that would explain the process? I'm always eager to learn new things! ;)

**Reply by dlech** [reply]
The simplest way would probably be to copy the `MailboxRFCOMMClient` class definition and add the channel as a parameter.

source code: https://github.com/pybricks/pybricks-micropython/blob/04c7304a25b8b16c6cd31ac1caaac501f3fb57be/bricks/ev3dev/modules/pybricks/messaging.py#L281-L290

**Reply by bradocchs** [reply]
Yep, that worked like a charm! Thanks so much.

**Reply by bradocchs** [reply]
Quick update: I know this is a hack and not recommended, but Python will also let you do this:
```
import pybricks.messaging as pm
pm.EV3_RFCOMM_CHANNEL = 5
```
"If it works..." ;)

**Comment by bradocchs** [comment]
> > PS - Not sure why my code is not formatting as code...
> 
> I fixed it up. The trick is to use a "code fence with 3 backticks like this:
> 
> ```
> ```python
> # code goes here
> ...
> ```
> ```

Thanks! I was just using the <> button on the toolbar.
